src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,118 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int ELEMENT_NOT_FOUND = -1;
+    private T[] elements;
+    private int size;
+
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index " + index + " out of bounds for size " + size);
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index " + index + " out of bounds for size " + size);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAdd(index);
+        ensureCapacity(size + 1);
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
```

Comment:
You can rewrite like this:
```
    @Override
    public T remove(int index) {
        checkIndex(index);
        checkIndexEqualsSize(index);
        T element = elements[index];
        System.arraycopy(elements, index + 1, elements, index, --size - index);
        return element;
    }
```

Reply:
> You can rewrite like this:
> 
> ```
>     @Override
>     public T remove(int index) {
>         checkIndex(index);
>         checkIndexEqualsSize(index);
>         T element = elements[index];
>         System.arraycopy(elements, index + 1, elements, index, --size - index);
>         return element;
>     }
> ```

I don't understand how to implement the checkIndexEqualsSize(index) method, I get two errors when passing tests
java.lang.AssertionError: Test failed! Returned value should be "Java", but was "null"
 
java.lang.Exception: Unexpected exception, expected<java.util.NoSuchElementException> but was<core.basesyntax.ArrayListIndexOutOfBoundsException>